### PR TITLE
Supporting changes for Cloud email invites

### PR DESCRIPTION
**Pull request by timothyb89**
This adds a number of changes to support email user invites for Cloud.

image

* Adds a new public `/preset-roles` endpoint that just returns the preset roles for the current version of Teleport. It's unauthenticated and intended for use in the new onboarding invitation UI. I don't think this has security implications  beyond what we already return in the `ping` replies, since those include the Teleport version.
* Includes role descriptions in the JSON responses sent to the UI
* Makes various tweaks to our `react-select` wrapper, and adds a Creatable variant for email entry.
* Adds a new `requiresEmailLike` validation rule, and adds more typing to the rules
* Adds support for providing an invite collaborators element to replace the usual "Create User" UI from teleport.e
* Adds support for showing the invite collaborators element during onboard, if a flag is set as a query parameter in the invite URL.

Note that gravitational/teleport.e#2286 is needed to test the changes. This PR must be merged before the `teleport.e` PR, then another PR here should bump the .e ref.

See also:
 * https://github.com/gravitational/cloud/issues/5137
 * .e changes: https://github.com/gravitational/teleport.e/pull/2286
 * Sales Center RPC: https://github.com/gravitational/cloud/pull/5970

### CODE REVIEW

File: web/packages/teleport/src/Welcome/NewCredentials/NewCredentials.tsx
Comment (kimlisa) [inline web/packages/teleport/src/Welcome/NewCredentials/NewCredentials.tsx:70]: i'd avoid naming things `setXXX` unless it's a react hook state setter (that's how it's normally named `setXXX`), not sure how to give rename suggestions b/c i don't really understand it yet

File: web/packages/teleport/src/Welcome/NewCredentials/types.ts
Comment (kimlisa) [inline web/packages/teleport/src/Welcome/NewCredentials/types.ts:72]: oh so it is a setter, i think setting a default to false was a mistake then? (see previous comment)

**Review by kimlisa — COMMENTED** [review]
first pass, will continue after a few questions:

- not sure if i'm doing something wrong (or misunderstanding something), but this these changes should only affect cloud users right? (i'm running enterprise and i'm seeing these changes, i can't add a user through the web ui)

- what if a user wants to add a non-email user?

- anyway you can spin a cluster with all your changes so i can test it? 🙏 


also there is a weird flash when rendering the dialog because of this height:

image

**Comment by timothyb89** [comment]
> first pass, will continue after a few questions:
> 
> * not sure if i'm doing something wrong (or misunderstanding something), but this these changes should only affect cloud users right? (i'm running enterprise and i'm seeing these changes, i can't add a user through the web ui)

Yes, it should only affect Cloud. I'll look into why you're seeing it on Enterprise.

> * what if a user wants to add a non-email user?

The decision was that non-email users can no longer be created on Cloud. If someone wants to add one, they'll need to use `tctl`. I think the plan is to see what sort of customer feedback we get on that change.

>  * anyway you can spin a cluster with all your changes so i can test it? 🙏

Hmm, that's a good idea, I think it'll be possible once the changes in Sales Center are deployed (hopefully soon): https://github.com/gravitational/cloud/pull/5970.

**Comment by timothyb89** [comment]
Various TODOs needed before this can be merged:

- [ ] Add unit tests
- [ ] Remove `e` ref bump - the changes are interdependent so the supporting changes will need to be merged sans the `e` bump, then we'll bump it via a separate PR.

File: web/packages/teleport/src/Users/Users.tsx
Comment (mcbattirola) [inline web/packages/teleport/src/Users/Users.tsx:54]: We can destructure `inviteCollaborators` here to follow the other props.

File: web/packages/teleport/src/Users/Users.tsx
Comment (mcbattirola) [inline web/packages/teleport/src/Users/Users.tsx:67]: If I'm reading correctly, this will display `Enroll Users` for any enterprise customers since we are always sending this on enterprise features (https://github.com/gravitational/teleport.e/blob/f32252614746014c5a9136cd34a9bf3ee2dc5d8f/web/teleport/src/features.tsx#L376)

You could use the `isCloud` flag on `web/packages/teleport/src/config.ts` to show/hide this instead.

**Review by michellescripts — APPROVED** [review]

File: web/packages/teleport/src/Users/Users.tsx
Comment (timothyb89) [reply web/packages/teleport/src/Users/Users.tsx:54]: Sure, I'd done it this way since it made the type checker angry, turns out intrinsic elements just need to be capitalized? :facepalm:  Fixed now at least.

File: web/packages/teleport/src/Users/Users.tsx
Comment (timothyb89) [reply web/packages/teleport/src/Users/Users.tsx:67]: Yup, that did the trick, thanks!

File: web/packages/teleport/src/Welcome/NewCredentials/NewCredentials.tsx
Comment (timothyb89) [reply web/packages/teleport/src/Welcome/NewCredentials/NewCredentials.tsx:70]: Right, sorry, I was following the pattern from `setDisplayOnboardingQuestionnaire` but I think I can just as well set it to `null`.

File: web/packages/teleport/src/Welcome/NewCredentials/types.ts
Comment (timothyb89) [reply web/packages/teleport/src/Welcome/NewCredentials/types.ts:72]: Yup, I'll set it to null by default instead (per the type signature I suppose).

File: web/packages/shared/components/FieldSelect/FieldSelect.tsx
Comment (timothyb89) [reply web/packages/shared/components/FieldSelect/FieldSelect.tsx]: Sure, I'll sync this with Cloud once it's merged.

**Comment by timothyb89** [comment]
A few non-threaded changes:
 * The flash of content on the dialog is fixed (the loading spinner no longer has a minimum height)
 * The Enroll Users button (and new feature) is no longer shown for Enterprise
 * The `e` ref update is removed, so for testing you'll need to check out https://github.com/gravitational/teleport.e/pull/2286

**Comment by timothyb89** [comment]
More notes: I think we've now got reasonable testing / storybook coverage for this PR (though I have some more work to do in the corresponding .e change).

Also, with the [cloud changes merged](https://github.com/gravitational/cloud/pull/5970), I've now got a test cluster running this change, so DM me if you'd like an account.

Additionally, per some feedback I've renamed the `?initial` parameter to `?invite` to make it what it does a little more explicit.

**Review by kimlisa — APPROVED** [review]
thanks a lot of the changes and the cluster 👍 

most of my comments are pretty minor so i'll approve this.

I think some of the feedback for below is actually being affected in enteprise:

some rule validation is kinda broken, it sometimes allows dups, sometimes it doesn't, and when it doesn't it duplicates error message

this is an error after making an api call (allows dups):

image

this prevents me, but the error msg's are duplciated:

image

nit:

image

probably caused by some kind of initial loading, i suggest trying out this css property: [visibility](https://developer.mozilla.org/en-US/docs/Web/CSS/visibility). you'd probably have to lay the loader on top of this block

**Review by mcbattirola — APPROVED** [review]
LGTM once Lisa's comments are addressed

File: web/packages/shared/components/Select/types.ts
Comment (timothyb89) [reply web/packages/shared/components/Select/types.ts:52]: It does do that, but it also lets us highlight items selectively. That's how specific invalid addresses get highlighted in red.

File: web/packages/shared/components/FieldSelect/types.tsx
Comment (timothyb89) [reply web/packages/shared/components/FieldSelect/types.tsx:1]: Given `defaultRule` is also not a type, I've just renamed this file to `shared.tsx`. It still has `LabelTip` which needs .tsx.

File: web/packages/shared/components/FieldSelect/types.tsx
Comment (timothyb89) [reply web/packages/shared/components/FieldSelect/types.tsx:23]: (duplicating the other comment) I'll rename the file outright since the other function is also not a type.

File: web/packages/shared/components/Select/Select.story.tsx
Comment (timothyb89) [reply web/packages/shared/components/Select/Select.story.tsx]: I've split it into its own dedicated story and added labels. It definitely works on my machine, though in this mode (without a custom `keyDown` handler like we use in `InviteCollaboratorsForm`) input is a bit goofy, just due to react-select's defaults. I added some comments to explain what you might want to do to actually use the component in a  real world setting.

**Comment by timothyb89** [comment]
> some rule validation is kinda broken, it sometimes allows dups, sometimes it doesn't, and when it doesn't it duplicates error message

I've added a new rule to ensure there's no duplicates within the entered field (we only checked against cluster users), and refactored the validation to aggregate all the various error messages so the strings aren't too verbose. I also noticed we didn't error highlight for user exists/duplicates, so added a new validator for that as well. 

It now looks like this (hopefully worst case):
image

(these changes are mostly in the .e PR)

> probably caused by some kind of initial loading, i suggest trying out this css property: [visibility](https://developer.mozilla.org/en-US/docs/Web/CSS/visibility). you'd probably have to lay the loader on top of this block

I've now set it to use `visibility: hidden` while loading (also in the .e PR).

File: lib/web/apiserver.go
Comment (zmb3) [inline lib/web/apiserver.go:770]: I don't think we typically put - characters in URL paths.

File: lib/web/resources.go
Comment (zmb3) [inline lib/web/resources.go:160]: Are we returning the full content of the role to unauthenticated users?

Comment (timothyb89) [reply lib/web/resources.go:160]: Yes, since this is used during onboarding (before credentials are created) as an alternative to fetching the actual roles on the cluster, which would reveal specific information and requires auth. For what it's worth, `auth.GetPresetRoles()` does not query the database - it is the same code that generates the roles that _would_ be inserted, and will always return the same static content for a given binary, regardless of any changes users make to those roles.

These are hard-coded and correspond directly to the Teleport version, which is already available to unauthenticated users via the `ping` endpoint, so an attacker could get the same information by cloning our git repo and checking out the appropriate tag. Hopefully if/when we decide to mask the Teleport version (and this new endpoint) we'll have revised the onboarding flow to be able to query for actual cluster roles.

File: lib/web/apiserver.go
Comment (timothyb89) [reply lib/web/apiserver.go:770]: I'll rename it to `presetroles`

File: lib/web/resources.go
Comment (zmb3) [reply lib/web/resources.go:160]: Thanks for the detail. Sounds fine to me.

**Review by zmb3 — APPROVED** [review]

**Pull request merged**
